src/stash/src/upgrade.rs

```diff
@@ -0,0 +1,53 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::BTreeMap;
+
+use super::InternalStashError;
+use crate::{StashError, Transaction, TypedCollection};
+
+pub enum MigrationAction<K1, K2, V2> {
+    /// Deletes the provided key.
+    Delete(K1),
+    /// Inserts the provided key-value pair. The key must not currently exist!
+    Insert(K2, V2),
+    /// Updates the provided key to the new value. The key must exist!
+    Update(K1, V2),
```

Comment:
Yeah I had one originally, but I thought it might be useful to have the separate "insert" vs "update" semantics, more than happy to add it in the future